pkg/xds/sync/ingress_proxy_builder.go

```diff
@@ -96,74 +77,68 @@ func (p *IngressProxyBuilder) buildZoneIngressProxy(
+	if err := p.ResManager.Get(ctx, zoneIngress, core_store.GetBy(key)); err != nil {
 		return nil, err
 	}
 	// Update Ingress' Available Services
 	// This was placed as an operation of DataplaneWatchdog out of the convenience.
 	// Consider moving to the outside of this component (follow the pattern of updating VIP outbounds)
-	if err := p.updateIngress(ctx, zoneIngress); err != nil {
+	if err := p.updateIngress(ctx, zoneIngress, meshContexts); err != nil {
 		return nil, err
 	}
 	return zoneIngress, nil
 }
 
-func (p *IngressProxyBuilder) updateIngress(ctx context.Context, zoneIngress *core_mesh.ZoneIngressResource) error {
-	allMeshDataplanes := &core_mesh.DataplaneResourceList{}
-	if err := p.ReadOnlyResManager.List(ctx, allMeshDataplanes); err != nil {
-		return err
-	}
-	allMeshGateways := &core_mesh.MeshGatewayResourceList{}
-	if err := p.ReadOnlyResManager.List(ctx, allMeshGateways); err != nil {
-		return err
-	}
-	allMeshDataplanes.Items = xds_topology.ResolveAddresses(syncLog, p.LookupIP, allMeshDataplanes.Items)
-
-	availableExternalServices, err := p.getIngressExternalServices(ctx)
-	if err != nil {
-		return err
-	}
-
+func (p *IngressProxyBuilder) updateIngress(
+	ctx context.Context, zoneIngress *core_mesh.ZoneIngressResource,
+	meshContexts xds_context.MeshContexts,
+) error {
 	// Update Ingress' Available Services
 	// This was placed as an operation of DataplaneWatchdog out of the convenience.
 	// Consider moving to the outside of this component (follow the pattern of updating VIP outbounds)
-	return ingress.UpdateAvailableServices(ctx, p.ResManager, zoneIngress, allMeshDataplanes.Items, allMeshGateways.Items, availableExternalServices.Items, p.ingressTagFilters)
+	return ingress.UpdateAvailableServices(
+		ctx,
+		p.ResManager,
+		zoneIngress,
+		meshContexts.AllDataplanes(),
+		meshContexts.AllMeshGateways(),
+		p.getIngressExternalServices(meshContexts).Items,
+		p.ingressTagFilters,
+	)
 }
 
-func (p *IngressProxyBuilder) getIngressExternalServices(ctx context.Context) (*core_mesh.ExternalServiceResourceList, error) {
-	meshList := &core_mesh.MeshResourceList{}
-	if err := p.ReadOnlyResManager.List(ctx, meshList, store.ListOrdered()); err != nil {
-		return nil, err
-	}
-
+func (p *IngressProxyBuilder) getIngressExternalServices(
+	meshContexts xds_context.MeshContexts,
+) *core_mesh.ExternalServiceResourceList {
 	allMeshExternalServices := &core_mesh.ExternalServiceResourceList{}
 	var externalServices []*core_mesh.ExternalServiceResource
-	for _, mesh := range meshList.Items {
+	for _, mesh := range meshContexts.Meshes {
 		if !mesh.ZoneEgressEnabled() {
 			continue
 		}
-		meshName := mesh.GetMeta().GetName()
 
-		meshCtx, err := p.meshCache.GetMeshContext(ctx, meshName)
-		if err != nil {
-			return nil, err
+		meshCtx, ok := meshContexts.MeshContextsByName[mesh.GetMeta().GetName()]
+		if !ok {
+			panic("there should be a corresponding mesh context for every mesh in mesh contexts")
```

Comment:
Maybe you can add a method to MeshContexts called `MustGetMesh` that panics when there is no mesh